{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../')\n",
    "from utility.file_utility import FileUtility\n",
    "from scipy.sparse import csr_matrix\n",
    "import os\n",
    "import tqdm\n",
    "path = '/mounts/data/proj/asgari/dissertation/datasets/deepbio/taxonomy/ncbi-blast-2.5.0+/bin/'\n",
    "os.environ['PATH'] += ':'+path\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from utility.math_utility import get_sym_kl_rows\n",
    "from nltk import FreqDist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class FastaMarkers2Excall:\n",
    "    \n",
    "    def __init__(self,fasta_file, matrix_path, feature_file_path, p_value_threshold=0.01):\n",
    "        self.seq_IDS=FileUtility.read_fasta_sequences_ids(fasta_file)\n",
    "        self.ez_taxa_dict={x.split()[0]:x.split()[1].split(';') for x in FileUtility.load_list('/mounts/data/proj/asgari/dissertation/git_repos/16S_datasets/EZ/raw/eztaxon_id_taxonomy.txt')}\n",
    "        self.align_markers(p_value_threshold)\n",
    "        \n",
    "        self.mat=FileUtility.load_sparse_csr(matrix_path)\n",
    "        self.features=FileUtility.load_list(feature_file_path)\n",
    "        self.redundant_columns_indentification()\n",
    "    \n",
    "    def extract_results():\n",
    "        results=[]\n",
    "        for group in self.equiv_classes:\n",
    "            results.append([self.aligned_markers[x] for x in group])\n",
    "        \n",
    "            \n",
    "    def update_matrix_by_markers(self):\n",
    "        '''\n",
    "            called by align_markers\n",
    "        '''\n",
    "        new_matrix=[]\n",
    "        for feature, taxnomy, pvalue in self.aligned_markers:\n",
    "            column=self.features.index(feature)\n",
    "            new_matrix.append(self.mat[:,column].toarray().T[0].tolist())\n",
    "            new_matrix=np.array(new_matrix)\n",
    "        self.update_matrix=new_matrix\n",
    "\n",
    "    @staticmethod\n",
    "    def find_equiv_classes(list_of_pairs):\n",
    "        found_list=[]\n",
    "        for x,y in list_of_pairs:\n",
    "            if found_list==[]:\n",
    "                found_list.append(set([x,y]))\n",
    "            else:\n",
    "                idx_to_add=-1\n",
    "                for idx,group in enumerate(found_list):\n",
    "                    if x in group or y in group:\n",
    "                        idx_to_add=idx\n",
    "                if idx_to_add==-1:\n",
    "                    found_list.append(set([x,y]))\n",
    "                else:\n",
    "                    found_list[idx_to_add]=found_list[idx_to_add].union(set([x,y]))\n",
    "        return found_list\n",
    "    \n",
    "    @staticmethod\n",
    "    def isGenomeName(inputString):\n",
    "    return np.sum([1 if char.isdigit() else 0 for char in inputString])/np.sum([1 if char.isalpha() else 0 for char in inputString])>0.8\n",
    "\n",
    "    def lowest_certain_level(self,results):\n",
    "        \n",
    "        levels_id={'Superkingdom':1,'phylum':1,'class':2,'order':3,'family':4,'genus':5,'species':6}\n",
    "        species=set([x[0][levels_id['species']] for x in results])\n",
    "        genuses=set([x[0][levels_id['genus']] for x in results])\n",
    "        families=set([x[0][levels_id['family']] for x in results])\n",
    "        orders=set([x[0][levels_id['order']] for x in results])\n",
    "        classes=set([x[0][levels_id['class']] for x in results])\n",
    "        phylums=set([x[0][levels_id['phylum']] for x in results])\n",
    "        Superkingdoms=set([x[0][levels_id['Superkingdom']] for x in results])\n",
    "\n",
    "        if len(species)==1:\n",
    "            return ';'.join(results[0][0])\n",
    "        elif len(genuses)==1:\n",
    "            return ';'.join(results[0][0][0:6])\n",
    "        elif len(families)==1:\n",
    "            return ';'.join(results[0][0][0:5])\n",
    "        if len(orders)==1:\n",
    "            return ';'.join(results[0][0][0:4])\n",
    "        elif len(classes)==1:\n",
    "            return ';'.join(results[0][0][0:3]) \n",
    "        elif len(phylums)==1:\n",
    "            return ';'.join(results[0][0][0:2]) \n",
    "        elif len(Superkingdoms)==1:\n",
    "            return ';'.join(results[0][0][0:1]) \n",
    "        else:\n",
    "            return False\n",
    "    \n",
    "    def redundant_columns_indentification(self):\n",
    "        distances=get_sym_kl_rows(self.update_matrix)\n",
    "        flatten_distances=distances.flatten()\n",
    "        list_of_pairs=np.argwhere(distances<np.percentile(d, 10, axis=0)).tolist()\n",
    "        self.equiv_classes=FastaMarkers2Excall.find_equiv_classes(list_of_pairs)\n",
    "        \n",
    "    def align_markers(self,p_value_threshold):\n",
    "        final_results=[]\n",
    "        for idx, (seq, description) in tqdm.tqdm(self.seq_IDS.items()):\n",
    "            pval=float(description.split(':')[1])\n",
    "            if pval<=p_value_threshold:\n",
    "                FileUtility.create_fasta_file('temp.fasta',[seq],['temp'])\n",
    "                blastx_cline=NcbiblastnCommandline(query='temp.fasta', db=\"/mounts/data/proj/asgari/dissertation/git_repos/16S_datasets/EZ/raw/eztaxon_qiime_full.fasta\", evalue=0.001, outfmt=5, out=\"temp.xml\")\n",
    "                blastx_cline()\n",
    "                f=open(\"temp.xml\",'r')\n",
    "                blast_records = NCBIXML.parse(f)\n",
    "                flag=False\n",
    "                score=-1\n",
    "                alignment_length=-1\n",
    "                results=[]\n",
    "                for blast_record in blast_records:\n",
    "                    for alignment in blast_record.alignments:\n",
    "                        for hsp in alignment.hsps:\n",
    "                            if not flag and score==-1:\n",
    "                                score=hsp.score\n",
    "                                alignment_length=hsp.align_length\n",
    "                                flag=True\n",
    "                            if hsp.score >= score and hsp.align_length>=alignment_length:\n",
    "                                results.append((self.ez_taxa_dict[alignment.hit_id],hsp.expect))\n",
    "                if len(results)>0:\n",
    "                    res=self.lowest_certain_level(results)\n",
    "                    if res:\n",
    "                        final_results.append((seq,res+idx[-1],pval))\n",
    "                else:\n",
    "                    final_results.append((seq,'ZZZNOVEL'+idx[-1],pval))\n",
    "        # sorted markers by the taxonomy information of the last certain level\n",
    "        self.aligned_markers=sorted(final_results, key=operator.itemgetter(1), reverse=False)\n",
    "        self.min_p_value=p_value_threshold\n",
    "        self.update_matrix_by_markers()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
